link-grammar/dict-common/regex-morph.c

```diff
@@ -48,84 +54,199 @@ typedef struct {
+	pcre2_get_error_message(rc, errbuf, ERRBUFFLEN);
+	prt_error("Error: Regex matching error: \"%s: %s\": %s (code %d)\n",
+	          rn->name, rn->pattern, errbuf, rc);
+	return false;
+}
+
+static void reg_free(Regex_node *rn)
+{
+	regex_t *re = rn->re;
+
+	pcre2_match_data_free(re->re_md);
+	pcre2_code_free(re->re_code);
+	free(re);
+}
+#endif // HAVE_PCRE2_H
+
+/* ========================================================================= */
+#if USE_CXXREGEX
+static bool reg_comp(Regex_node *rn)
+{
+	try
+	{
+		rn->re = new std::regex(rn->pattern);
+	}
+	catch (const std::regex_error& e)
+	{
+		prt_error("Error: Failed to compile regex '%s' (%s): %s (code %d)",
+		          rn->pattern, rn->name, e.what(), e.code());
+		return false;
+	}
+
+	return true;
+}
+
+static bool reg_match(const char *s, const Regex_node *rn)
+{
+	bool match;
+
+	try
+	{
+		match = std::regex_search(s, *(std::regex*)rn->re);
+	}
+	catch (const std::regex_error& e)
+	{
+		prt_error("Error: Regex matching error '%s' (%s): %s (code %d)",
+		          rn->pattern, rn->name, e.what(), e.code());
+		match = false;
+	}
+
+	return match;
+}
 
-	prt_error("Error: %s: \"%s\" (%s", msg, re->pattern, re->name);
-	if (-1 != erroffset) prt_error(" at %d", erroffset);
-	prt_error("): %s (%d)\n", errbuf, rc);
+static void reg_free(Regex_node *rn)
+{
+	delete (std::regex *)rn->re;
 }
+#endif // USE_CXXREGEX
+
+/* ============================== Internal API ============================= */
 
 /**
  * Compile all the given regexs.
- * Return 0 on success, else an error code.
+ * Return \c true on success, else \c false.
  */
-int compile_regexs(Regex_node *rn, Dictionary dict)
+bool compile_regexs(Regex_node *rn, Dictionary dict)
 {
 	while (rn != NULL)
 	{
 		/* If rn->re non-null, assume compiled already. */
 		if(rn->re == NULL)
 		{
-			int rc;
-			regex_t *re = rn->re = malloc(sizeof(regex_t));
-
-#if HAVE_PCRE2_H
-			PCRE2_SIZE erroffset;
-			re->re_code =
-				pcre2_compile((PCRE2_SPTR)rn->pattern, PCRE2_ZERO_TERMINATED,
-				              PCRE2_UTF|PCRE2_UCP, &rc, &erroffset, NULL);
-			if (NULL != re->re_code)
-			{
-				rc = 0;
-				re->re_md = pcre2_match_data_create(0, NULL);
-				if (NULL == re->re_md) return -1; /* Unhandled for now. */
-			}
-#else
-			const int erroffset = -1;
-
-			/* REG_ENHANCED is needed for macOS to support \w etc. */
-#ifndef REG_ENHANCED
-#define REG_ENHANCED 0
-#endif
-			/* REG_GNU is needed for NetBSD to support \w etc. */
-#ifndef REG_GNU
-#define REG_GNU 0
-#endif
-			rc = regcomp(re, rn->pattern,
-			             REG_NOSUB|REG_EXTENDED|REG_ENHANCED|REG_GNU);
-#endif
-
-			if (rc)
+			if (!reg_comp(rn))
 			{
-				prt_regerror("Failed to compile regex", rn, rc ,erroffset);
 				rn->re = NULL;
-				return rc;
+				return false;
 			}
 
 			/* Check that the regex name is defined in the dictionary. */
-			if ((NULL != dict) && !dict_has_word(dict, rn->name))
+			if ((dict != NULL) && !dict_has_word(dict, rn->name))
```

Comment:
The reason for writing `NULL != dict` instead of `dict != NULL` is to help avoid "undetectable single-byte errors". In this case, if you accidentally delete the `!` the first will not compile; the second will compile but produce incorrect results. 

These kinds of errors are very rare an unusual, but I did watch one guy spend a week debugging something with a single-byte error.  Then again, he was a bit of a, umm. ahh. Actually, I think he was stoned all the time.